pkg/controller/drain/drain_controller.go

```diff
@@ -0,0 +1,417 @@
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing node %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(2).Infof("Dropping node %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+func (ctrl *Controller) syncNode(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing node %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing node %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// First check if a drain request is needed for this node
+	// TODO (jerzhang): consider not using the cache for this,
+	// since it has been shown to be problematic in the past
+	node, err := ctrl.nodeLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("node %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	desiredState := node.Annotations[daemonconsts.DesiredDrainerAnnotationKey]
+	if desiredState == node.Annotations[daemonconsts.LastAppliedDrainerAnnotationKey] {
+		glog.V(4).Infof("Node %v has the correct drain", key)
+		return nil
+	}
+
+	podName, ok := os.LookupEnv("POD_NAME")
+	if !ok || podName == "" {
+		glog.Fatalf("pod name is required for controller drainer")
+	}
+
+	drainer := &drain.Helper{
+		Client:              ctrl.kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteEmptyDirData:  true,
+		GracePeriodSeconds:  -1,
+		Timeout:             90 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			glog.Infof("%s pod %s/%s", verbStr, pod.Namespace, pod.Name)
+		},
+		AdditionalFilters: []drain.PodFilter{
+			func(pod corev1.Pod) drain.PodDeleteStatus {
+				if pod.Name == podName {
+					return drain.MakePodDeleteStatusSkip()
+				}
+				return drain.MakePodDeleteStatusOkay()
+			},
+		},
+		Out:    writer{glog.Info},
+		ErrOut: writer{glog.Error},
+		Ctx:    context.TODO(),
+	}
+
+	desiredVerb := strings.Split(desiredState, "-")[0]
+	if desiredVerb == daemonconsts.DrainerStateUncordon {
+		glog.Infof("node %v to be uncordoned", key)
+		// perform uncordon
+		if err := ctrl.cordonOrUncordonNode(false, node, drainer); err != nil {
+			return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
+		}
+	}
+
+	if desiredVerb == daemonconsts.DrainerStateDrain {
+		// First check if we have an ongoing drain
+		// This is currently stored in the object itself as a map but,
+		// Practically during upgrades the control plane node this controller
+		// pod is running on will also be terminated (the drainer will skip it).
+		// This is a bit problematic in practice since we don't really have a previous state.
+		// TODO (jerzhang) consider using a new CRD for coordination
```

Comment:
Hmm.  What if the Node object had an ObjectReference to the thing doing the drain?  The benefit of that is it'd become a clear central point of knowledge/coordination.

Short term maybe we could shoehorn such a reference into an annotation.

For example, suppose that this drain pod is killed unexpectedly (control plane node went down, it hit OOM or whatever).

The pod restarts on another host (or same host in OOM scenario after OOM clears out), sees the annotation, and checks whether the pod that it references is still active or not.

If not, it knows to reset the annotation.  Otherwise, we should skip trying to drain and assume the other pod has it?

Reply:
As in, so long as we started the drain and got to the eviction API call, it's technically always ongoing.

Reply:
Or to rephrase, while in Go we're making *one* API call, it internally is implemented with many requests to the API server (see `evictPods()`) and hence is not transactional, so we need to handle being interrupted in the middle and restarting.

Reply:
OK.  Seems fixable by serializing the start time into another annotation or so, but not a blocker.

This all said, I think this "drain killing itself while also draining other nodes" is definitely something to exercise explicitly in manual (and ideally automated tests).

I realized it's simple to do via e.g.:

- Create MC for worker pool
- `oc -n openshift-machine-config-operator delete pod/machine-config-controller-xyz`

(Of course, this is a more synthetic version of the real flow of actually rolling the master pool, but I think should hit the relevant edge cases while being a lot cheaper to test and easier to precisely time the termination, etc.)

Reply:
I think another good test to exercise alongside this would be to have an in-progress update on the worker pool, start a second update, delete said pod, and see if the new controller recovers.

And, hmm, maybe recovery from degrade testing, and timeout behaviour, although those are maybe harder to do in a e2e context.

If we can confirm all 3 works, I think we should be good to go in terms of feature parity with what we have (minus some additional slowdowns)